docs/enhancements/replace_broken_clusters.md

```diff
@@ -0,0 +1,69 @@
+# Automatically detect and replace broken Clusters
+
+## Summary
+
+Hibernating clusters often fail to transition to running state, but are assigned to ClusterClaims today. We would want to 
+validate a cluster can be successfully resumed and is running as expected before assigning it to a claim.
+
+## Motivation
+
+As we move towards being more proactive with having clusters ready as [hot spares](https://github.com/openshift/hive/pull/1434),
+we could come across instances where stale hibernated clusters could fail to resume and still be assigned to ClusterClaims.
+There's a need to detect broken clusters and replace them automatically.
+
+### Goals
+
+Always strive to assign a running cluster to a claim
+
+### Non Goals
+
+Fixing clusters (or attached claims) once a running cluster has already been assigned.
+
+## Proposal
+
+### Stand-by clusters
+
+Currently, we consider all installed clusters which are not set to be deleted as Assignable clusters. 
+Introduce another bucket of stand by clusters, which are installed and are not set for deletion, but are hibernating.
+A cluster is only assignable if it is running. A hibernating cluster has to be resumed successfully to be added to the 
+assignable clusters bucket.
+Some of the factors that might inhibit successful resume of a cluster that has been hibernating for a while include  
+unapproved certificates and expired credentials. Introduction of a separate bucket for hibernating clusters 
+allows for a validation before it is assigned to a claim.
+
+### Transition to Assignable clusters
+
+If a claim is created and there are no assignable clusters, we would resume a hibernating cluster and check if the resume
+was successful. The validating condition could be the expected powerstate matches the actual powerstate. If we encounter
+an error, we can mark that cluster to be deleted and try to transition the next available stand-by cluster.
+Figuring out how long we wait for the resume process to complete can be tricky - as there are instances when we can wait
+indefinitely. So we can consider a timeout for that wait. Fine-tuning the said timeout would be dependent on the 
+platform and the infrastructure, hence left for the admins to decide. Default is no time-out - so waiting indefinitely 
+unless an error is encountered.
+
+#### Timeout in Hibernation controller
+
+Introduce an optional spec.powerstateTransitionTimeout in cluster deployments. Once the powerstate is changed and 
```

Comment:
I wonder if it makes sense to have separate timeouts for stopping vs resuming. Do/can those take significantly different amounts of time?

Also:
```suggestion
Introduce an optional spec.powerstateTransitionTimeout in cluster deployments, and a corresponding field in ClusterPool. Once the powerstate is changed and 
```

Reply:
I think that's a good idea. I was testing for https://issues.redhat.com/browse/HIVE-1630 and observed that resume time is significantly greater than stop time.
![image](https://user-images.githubusercontent.com/54181650/130511754-d41a1b61-5040-420f-b8f9-aecc81992ead.png)
My cluster was fresh, I am assuming a stale cluster (or a cluster that has been hibernating for more than 24 hours) would take longer to resume

Reply:
Do we need it on ClusterDeployment at all? It seems like it could just live on Pool and handle timeouts there. Is there any value exposing it on standalone cluster deployments?

Reply:
> value exposing it on standalone cluster deployments?

The thing we're timing out is the Hibernating condition of the CD. It makes sense to do that via the hibernation controller, which doesn't have any pool logic at the moment.

But aside from code organization reasons, doesn't this seem like a generally useful knob to have on CDs? Today a consumer resuming a CD (whether or not it came from a pool) need some higher-level logic to keep time if they want to flag "this is taking too long". With this change, we would be giving them a simple condition reason to watch for instead.

Reply:
Ah I thought we already did a timeout condition reason for this, that may not be the case. Making it configurable amount of time, maybe worthwhile. If so then lets put a sub-struct for power state on the list for v2 API changes. :)

Reply:
> I thought we already did a timeout condition reason for this

We'll wait to kick off hibernation until syncsets have been applied; but [only for 10m, after which we'll hibernate anyway](https://github.com/openshift/hive/blob/6f060f04b8ab4fb505b20486b09bba54c631974e/pkg/controller/hibernation/hibernation_controller.go#L203-L210). Other than that and `hibernateAfter` logic, I don't see anything around timeouts. (At least not in the hibernation controller. No idea what's happening on the cloud -- except that we've observed "stuck resuming" lasting indefinitely, so...)

> sub-struct for power state on the list for v2 API changes.

Something like this:

```
spec:
  ...
  powerState:
    desiredPowerState: {Running|Hibernating}
    runningTimeout: [Duration]
    hibernatingTimeout: [Duration]
  ...
```
?